### ISSUE TITLE
Decrypt early data before sending the servers early_data_extension?

### ISSUE THREAD

**Issue by Tamec1** [openssl/openssl#20485]
When calling SSL_read_early_data(): 
-> the client hello is processed
-> The server hello is constructed 
-> if there is an early_data_extension and a max_early_data parameter > 0
{
-> the servers early_data_state is set to accepting
-> an early_data_extension (**indicating early_data will be accepted** by the server) is appended to the ServerHello
}
-> the ServerHello is send
-> **THEN** any early_data in the buffer is read, decrypted and handed over to the application 
     (returning from the called SSL_read_early_data() method)

With the above procedure it is not possible for the Client to check the unencrypted early data for anything unwanted
and thereafter decide on sending a server early_data_extension or not.
The Server is only able to completely deny early data by setting its max_early_data parameter to 0.

The main issue I have with this behaviour is:
When the server accepted early data and the client calls the method SSL_get_early_data_status()
it will return SSL_EARLY_DATA_ACCEPTED independent of the fact that the application on the server side 
really accepted and proccessed the data.
This is fine if the Client is aware that the early_data_status does not guarantee that the server side application is dealing
with the data and that there may still be the need to send it again after the handshake.

Long story short, two questions:
Is this the intended behaviour?
Would it be more accurate to be able to first read the early data and then decide to send an extension or not?

I am using OpenSSL 3.0.8

Thank you very much in advance.

kind regards
Tamec

**Comment by davidben** [comment]
This is expressly forbidden by RFC 8446 and would cause a deadlock in clients that stream early data.
https://www.rfc-editor.org/rfc/rfc8446#section-4.2.11.3

If the server wishes to more selectively handle early data, it has a couple of choices at the application level. You can see RFC 8470 for how this is done in HTTP:
https://www.rfc-editor.org/rfc/rfc8470.html#section-3

Basically, you can either selectively buffer data until you've passed the 0-RTT point, or you can request a retry with an application-level mechanism. In HTTP, they added a new response code.

**Comment by Tamec1** [comment]
Thank you very much for your quick reply.
Somehow I missed that part in rfc 8446, thanks for pointing it out.
After looking at rfc 8470, it makes sense to me now 
to just buffer the incoming data and to take futher actions at the application level.

**Comment by mattcaswell** [comment]
Closing since this is marked as resolved.
